core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -5736,6 +5792,93 @@ private void rewriteSingleFirstLastGroupBy(QueryModel model) {
+
+            {
+                // Something like this: SELECT timestamp, * FROM trades LIMIT -3;
+                // compiles to:
+                // select-choose timestamp, symbol, side, price, amount, timestamp timestamp1
+                // from (select-choose timestamp, symbol, side, price, amount
+                // from (trades timestamp (timestamp))) limit -(3)
+                // We have this extra select choose to alias the already selected columns.
+                // We may be able to compact this
+
+                QueryModel nested = model.getNestedModel();
+                QueryModel none = nested.getNestedModel();
+
+                if (model.getLimitLo() != null
+                        && Chars.equals(model.getLimitLo().token, "-")
+                        && model.getLimitHi() == null
+                        && model.getUnionModel() == null
+                        && model.getJoinModels().size() == 1
+                        && nested.getJoinModels().size() == 1
+                        && none.getJoinModels().size() == 1
+                        && model.getGroupBy().size() == 0
+                        && nested.getGroupBy().size() == 0
+                        && none.getGroupBy().size() == 0
+                        && model.getSampleBy() == null
+                        && nested.getSampleBy() == null
+                        && !hasAggregateQueryColumn(model)
+                        && !model.isDistinct()) {
+
+                    IntList copiedColumns = new IntList();
+
+                    // we are going to compact the models
+                    if (model.getBottomUpColumns().size() >= nested.getBottomUpColumns().size()) {
+                        // since it is possible that this is a redundant model, or we can push down the extra select choose
+                        for (int i = 0, n = nested.getBottomUpColumns().size(); i < n; i++) {
+                            QueryColumn innerColumn = nested.getBottomUpColumns().get(i);
+
+                            if (!model.getColumnNameToAliasMap().contains(innerColumn.getName())) {
+                                // if it is not part of it, then is it the same as another column?
+                                CharSequence token = innerColumn.getAst().token;
+
+                                if (model.getColumnNameToAliasMap().contains(token)) {
+                                    // we can push down this column
+                                    nested.addBottomUpColumn(innerColumn);
+                                    copiedColumns.add(nested.getBottomUpColumns().size() - 1);
+                                }
+                            }
+                        }
+                    }
+
+                    boolean identicalModels = true;
+
+                    // now check that the models are identical
+                    if (model.getBottomUpColumns().size() == nested.getBottomUpColumns().size()) {
```

Comment:
What happens when `model.getBottomUpColumns().size() != nested.getBottomUpColumns().size()`? We end up with `identicalModels` set to true, so we get rid of the nested model. Is that the desired behavior?